bundles/org.openhab.core.crypto.cryptostore/src/main/java/org/openhab/core/crypto/cryptostore/CryptoStore.java

```diff
@@ -0,0 +1,328 @@
+
+    private int keySize = 128;
+    private String privKey = "";
+    private String cert = "";
+    private String caCert = "";
+    private String keystoreFileName = "";
+    private String keystoreAlgorithm = "RSA";
+    private int keyLength = 2048;
+    private String alias = "openhab";
+    private String caAlias = "trustedCa";
+    private String distName = "CN=openHAB, O=openHAB, L=None, ST=None, C=None";
+    private String cipher = "AES/GCM/NoPadding";
+    private String keyAlgorithm = "";
+
+    private @Nullable Cipher encryptionCipher;
+
+    public CryptoStore() {
+        try {
+            encryptionCipher = Cipher.getInstance(cipher);
+        } catch (NoSuchAlgorithmException | NoSuchPaddingException e) {
+            logger.debug("Could not get cipher instance", e);
+        }
+    }
+
+    public byte[] generateEncryptionKey() {
+        Key key;
+        try {
+            KeyGenerator keyGenerator = KeyGenerator.getInstance("AES");
+            keyGenerator.init(keySize);
+            key = keyGenerator.generateKey();
+            byte[] newKey = key.getEncoded();
+            this.keyAlgorithm = key.getAlgorithm();
+            return newKey;
+        } catch (NoSuchAlgorithmException e) {
+            logger.debug("Could not generate encryption keys", e);
+        }
+        return new byte[0];
+    }
+
+    private Key convertByteToKey(byte[] keyString) {
+        Key key = new SecretKeySpec(keyString, keyAlgorithm);
+        return key;
+    }
+
+    public String encrypt(String data, Key key) throws Exception {
+        return encrypt(data, key, this.cipher);
+    }
+
+    public String encrypt(String data, Key key, String cipher) throws Exception {
+        byte[] dataInBytes = data.getBytes();
+        Cipher encryptionCipher = this.encryptionCipher;
+        if (encryptionCipher != null) {
+            encryptionCipher.init(Cipher.ENCRYPT_MODE, key);
+            byte[] encryptedBytes = encryptionCipher.doFinal(dataInBytes);
+            return Base64.getEncoder().encodeToString(encryptedBytes);
+        } else {
+            return "";
+        }
+    }
+
+    public String decrypt(String encryptedData, Key key) throws Exception {
+        return decrypt(encryptedData, key, this.cipher);
+    }
+
+    public String decrypt(String encryptedData, Key key, String cipher) throws Exception {
+        byte[] dataInBytes = Base64.getDecoder().decode(encryptedData);
+        Cipher decryptionCipher = Cipher.getInstance(cipher);
+        Cipher encryptionCipher = this.encryptionCipher;
+        if (encryptionCipher != null) {
+            GCMParameterSpec spec = new GCMParameterSpec(keySize, encryptionCipher.getIV());
+            decryptionCipher.init(Cipher.DECRYPT_MODE, key, spec);
+            byte[] decryptedBytes = decryptionCipher.doFinal(dataInBytes);
+            return new String(decryptedBytes);
+        } else {
+            return "";
+        }
+    }
```

Comment:
I think it's a bit confusing to have a CryptoStore for managing public/private keys and certificates, but at the same time it has encrypt/decrypt functions using symmetric encryption. I don't really see the need for encrypting the keys in-memory (what threat is it supposed to protect against?), it just adds complexity for the binding developers.

Reply:
Yes but no matter how many keys you use to encrypt other keys, one must be stored somewhere in plaintext in memory (unless you want to ask the user to input it every time the private key is needed) so if someone is in a position to retrieve the private key, they can just as easily get the symmetric key and then decrypt it. 

I think your basic idea of a central warehouse is a good one, but as you say, cryptography is very hard to get right. I would be happy to continue discussing how it can be implemented in a good way.

Reply:
Agreed, and it's a step along the path.  Long term goal is to utilize things like TPMs to store keys.  I believe it will be a tiered mechanism that the user will ultimately decide.  For example, we could (for basic users) have them store a password in a file that's read on start.  For more advanced, we could query the user for that password on start to unlock the system.  For expert users, maybe we go with TPMs to store everything.

Reply:
But what is the actual threat here? Who is the attacker that would benefit from retrieving the key? What would they gain? The ability to control your TV? And if they can access memory on your OH server, they could probably do much more damage than that without using the key. Don't get me wrong, having a central credential storage would be a great addition to OH, but the cost of added complexity needs to be weighed against the actual risk.

Reply:
After reading a bit more on this, the KeyStore already is an implementation to store the keys securely (both in memory and on disk), so instead of implementing your own security solution, just retrieve the private key and certificate(s) from the KeyStore when needed and don't store it in your class' variables.

Reply:
I agree with pacive that encrypting keys in memory adds unnecessary complexity.  It's a nice goal, but it's a chicken and egg situation.  At some point, something has to be in cleartext which will unlock everything else.  The key isolation goal a very good one.

Reply:
The concern here was a "rogue binding" more than anything else.  For example, a marketplace binding.  We didn't want to risk that binding being able to get a hold of credentials or something like that.  Also, this is very much for the warehouse than really for this part of the implementation.

Reply:
Isn't that solved in an equally secure way by the password on the KeyStore? Also, @J-N-K had a proposal for isolating bindings from each other in [this comment](https://github.com/openhab/openhab-core/issues/3289#issuecomment-1368783070), which wouldn't complicate things for binding developers.